internal/bootstrap/deployer.go

```diff
@@ -425,17 +444,31 @@ func addLocalControllerCharm(ctx context.Context, objectStore services.Storage,
 		Name:     name,
 		Revision: archive.Revision(),
 	}
-	curl, err = uploader.PrepareLocalCharmUpload(curl.String())
+	curl, err = b.charmUploader.PrepareLocalCharmUpload(curl.String())
 	if err != nil {
 		return nil, nil, errors.Trace(err)
 	}
 
 	// Now we need to repackage it with the reserved URL, upload it to
 	// provider storage and update the state.
-	_, _, _, _, err = apiserver.RepackageAndUploadCharm(ctx, objectStore, uploader, archive, curl.String(), archive.Revision())
+	charmStorage := services.NewCharmStorage(services.CharmStorageConfig{
+		Logger:       b.logger,
+		StateBackend: b.charmUploader,
+		ObjectStore:  b.objectStore,
+	})
+
+	_, err = charmStorage.Store(ctx, curl.String(), downloader.DownloadedCharm{
+		Charm:        archive,
+		CharmData:    &charmArchiveBuf,
+		CharmVersion: archive.Version(),
+		Size:         int64(charmArchiveBuf.Len()),
+		SHA256:       archiveSHA256,
+		LXDProfile:   archive.LXDProfile(),
```

Comment:
And this wasn't required either.

Reply:
Resolved by rebasing on https://github.com/juju/juju/pull/18529